From 33fa020bd0658fcad42e083ff6a33bef63444871 Mon Sep 17 00:00:00 2001
From: Zhang Ning <ning.a.zhang@intel.com>
Date: Mon, 19 Sep 2016 15:14:40 +0800
Subject: [PATCH 1203/1214] rootfs: use async to populate rootfs

this patch is originally from:
https://lists.ubuntu.com/archives/kernel-team/2009-December/007544.html
with modify to fit new kernel, and less build warning.

it can save 160ms on boot, and independs on bootimage size.

Change-Id: I45640a61023f0b8a05e54a484d20671b45602ca8
Signed-off-by: Zhang Ning <ning.a.zhang@intel.com>
---
 include/linux/init.h |  2 ++
 init/initramfs.c     | 12 +++++++++++-
 init/main.c          |  8 ++++++++
 3 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/include/linux/init.h b/include/linux/init.h
index 586dd18..5b20299 100644
--- a/include/linux/init.h
+++ b/include/linux/init.h
@@ -260,6 +260,8 @@ extern bool initcall_debug;
 #define console_initcall(fn)	___define_initcall(fn, con, .con_initcall)
 #define security_initcall(fn)	___define_initcall(fn, security, .security_initcall)
 
+extern struct async_domain populate_rootfs_domain;
+
 struct obs_kernel_param {
 	const char *str;
 	int (*setup_func)(char *);
diff --git a/init/initramfs.c b/init/initramfs.c
index dab8d634..2ec670f 100644
--- a/init/initramfs.c
+++ b/init/initramfs.c
@@ -10,6 +10,7 @@
 #include <linux/syscalls.h>
 #include <linux/utime.h>
 #include <linux/file.h>
+#include <linux/async.h>
 
 static ssize_t __init xwrite(int fd, const char *p, size_t count)
 {
@@ -599,7 +600,9 @@ static void __init clean_rootfs(void)
 }
 #endif
 
-static int __init populate_rootfs(void)
+ASYNC_DOMAIN(populate_rootfs_domain);
+
+static void __init async_populate_rootfs(void *data, async_cookie_t cookie)
 {
 	/* Load the built in initramfs */
 	char *err = unpack_to_rootfs(__initramfs_start, __initramfs_size);
@@ -650,6 +653,13 @@ static int __init populate_rootfs(void)
 	 */
 	load_default_modules();
 
+	return;
+}
+
+static int __init populate_rootfs(void)
+{
+	async_schedule_domain(async_populate_rootfs, NULL, &populate_rootfs_domain);
 	return 0;
 }
+
 rootfs_initcall(populate_rootfs);
diff --git a/init/main.c b/init/main.c
index 4045566..8d9a111 100644
--- a/init/main.c
+++ b/init/main.c
@@ -1184,6 +1184,14 @@ static noinline void __init kernel_init_freeable(void)
 
 	(void) ksys_dup(0);
 	(void) ksys_dup(0);
+
+#ifdef CONFIG_BLK_DEV_INITRD
+	/*
+	 * We need to ensure that the filesystem is ready by this point,
+	 *  wait for async_populate_rootfs to complete.
+	 */
+	async_synchronize_full_domain(&populate_rootfs_domain);
+#endif
 	/*
 	 * check if there is an early userspace init.  If yes, let it do all
 	 * the work
-- 
2.7.4

